### PR TITLE
[`gpt`] Gpt2 fix half precision causal mask

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Applies a similar fix than https://github.com/huggingface/transformers/issues/23136 but for GPT2.

To reproduce: 
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("gpt2", device_map="auto", load_in_8bit=True)
inputs = torch.LongTensor([[1, 1, 1], [1, 2, 1]]).to(0)

print(model(inputs))
```

The explanation is the same as the tagged PR: 

> When going for low_cpu_mem_usage each parameter is force-casted to the expected dtype, which is force-set to torch.float16 for 8bit models.

> Therefore, for 8bit models (and also half-precision models) the causal mask is always force casted to float16 as it is part of the model's state dict, hence expected to be loaded from the Hub if the mask is available on the state dict.

> The fix is to add persistant=False and add a field _keys_to_ignore_on_unexpected (for removing the warnings) to avoid loading that causal mask from the state dict and assign it to the buffer, and all causal masks that are saved as buffers should do the same to avoid unexpected behaviors.

Some users reported that they were also able to reproduce on PyTorch main branch but without load_in_8bit, I didn't managed to reproduce that way, I will have a deeper look

cc @amyeroberts

### CODE REVIEW

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py [hunk 1]
```diff
@@ -746,7 +747,8 @@ class DecisionTransformerPreTrainedModel(PreTrainedModel):
     base_model_prefix = "decision_transformer"
     main_input_name = "states"
     supports_gradient_checkpointing = False
-    _keys_to_ignore_on_load_missing = [r"position_ids"]
+    _keys_to_ignore_on_load_missing = [r"position_ids", r"h\.\d+\.attn\.masked_bias", r"h\.\d+\.attn\.bias"]
```
Comment (amyeroberts) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:750]: Should `r"h\.\d+\.attn\.masked_bias", r"h\.\d+\.attn\.bias"` be in ` _keys_to_ignore_on_load_missing`?

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! I just have a question about `_keys_to_ignore_on_load_missing` for decision transformer

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for fixing :)

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for fixing :)
